packages/graphql-client/src/graphql-client/graphql-client.ts

```diff
@@ -165,3 +184,276 @@ function generateRequest(
+              .filter((chunk) => chunk.trim().length > 0)
+              .map((chunk) => {
+                const body = chunk
+                  .slice(
+                    chunk.indexOf(HEADER_SEPARATOR) + HEADER_SEPARATOR.length,
+                  )
+                  .trim();
+                return body;
+              });
+
+            if (chunkBodies.length > 0) {
+              yield chunkBodies;
+            }
+
+            buffer = buffer.slice(lastBoundaryIndex + boundary.length);
+
+            if (buffer.trim() === `--`) {
+              buffer = "";
+            }
+          }
+        }
+      } catch (error) {
+        throw new Error(
+          `Error occured while processing stream payload - ${getErrorMessage(
+            error,
+          )}`,
+        );
+      }
+    },
+  };
+}
+
+function generateRequestStream(
+  fetch: ReturnType<typeof generateFetch>,
+): GraphQLClient["requestStream"] {
+  return async (...props) => {
+    if (!DEFER_OPERATION_REGEX.test(props[0])) {
+      throw new Error(
+        formatErrorMessage(
+          "This operation does not result in a streamable response - use request() instead.",
+        ),
+      );
+    }
+
+    try {
+      const response = await fetch(...props);
+
+      const { status, statusText } = response;
+
+      if (!response.ok) {
+        throw new Error(statusText, { cause: response });
+      }
+
+      const responseContentType = response.headers.get("content-type") || "";
+      const isNotSupportedContentType = Object.values(CONTENT_TYPES).every(
+        (type) => !responseContentType.includes(type),
+      );
+
+      if (isNotSupportedContentType) {
+        throw new Error(
+          `${UNEXPECTED_CONTENT_TYPE_ERROR} ${responseContentType}`,
+          { cause: response },
+        );
+      }
+
+      if (responseContentType.includes(CONTENT_TYPES.json)) {
+        return {
+          async *[Symbol.asyncIterator]() {
+            const processedResponse = await processJSONResponse(response);
+
+            yield {
+              ...processedResponse,
+              hasNext: false,
+            };
+          },
+        };
+      }
+
+      const boundaryHeader = (responseContentType ?? "").match(
+        BOUNDARY_HEADER_REGEX,
+      );
+      const boundary = `--${boundaryHeader ? boundaryHeader[1] : "-"}`;
+
+      if (
+        !response.body?.getReader &&
+        !(response.body as any)![Symbol.asyncIterator]
+      ) {
+        throw new Error(
+          "API multipart response did not return an iterable body",
+          { cause: response },
+        );
+      }
+
+      const streamBodyIterator = getStreamBodyIterator(response);
+
+      let combinedData: Record<string, any> = {};
+      let responseExtensions: Record<string, any> | undefined;
+
+      return {
+        async *[Symbol.asyncIterator]() {
+          try {
+            let streamHasNext = true;
+
+            for await (const chunkBodies of readStreamChunk(
```

Comment:
I almost wonder if we want to extract this logic into its own object or class, like `StreamedResponse` or some such, so we can detach the client code from the parsing logic, in case we ever want to reuse it / for readability.

Reply:
We can - I kept them together because it was more easy to reason around when I was developing the feature. If it's easier to read to split them up, I can...but most likely, I'll keep the code in the same file since I don't know if I want to test the helper methods on their own 🤔

Reply:
From a reusability perspective, I honestly don't really think we will but if we do, we can abstract it out into it's own file then

Reply:
Yeah I think I'm ok with refactoring, but not necessarily into a whole new file. It's just a fairly long method and it's doing a few things, so it takes a second to fully understand it.